zap/src/parser/convert.rs

```diff
@@ -166,228 +204,326 @@ impl<'src> SyntaxEvDecl<'src> {
-						});
-					}
+				if let Ty::Opt(_) = parsed_ty {
+					self.report(Report::AnalyzeInvalidOptionalType {
+						span: (ty.span().end - 1)..ty.span().end,
+					});
 				}
 
-				Ty::Opt(Box::new(ty))
+				Ty::Opt(Box::new(parsed_ty))
 			}
 
-			SyntaxTyKind::Ref(name) => match name.into_config() {
-				"boolean" => Ty::Boolean,
-				"Color3" => Ty::Color3,
-				"Vector3" => Ty::Vector3,
-				"AlignedCFrame" => Ty::AlignedCFrame,
-				"CFrame" => Ty::CFrame,
-				"unknown" => Ty::Opt(Box::new(Ty::Unknown)),
-
-				_ => {
-					let name = name.into_config();
-
-					if !state.tydecl_exists(name) {
-						state.push_report(Report::AnalyzeUnknownTypeRef {
-							span: self.span(),
-							name,
-						});
+			SyntaxTyKind::Ref(ref_ty) => {
+				let name = ref_ty.name;
+
+				match name {
+					"boolean" => Ty::Boolean,
+					"Color3" => Ty::Color3,
+					"Vector3" => Ty::Vector3,
+					"AlignedCFrame" => Ty::AlignedCFrame,
+					"CFrame" => Ty::CFrame,
+					"unknown" => Ty::Opt(Box::new(Ty::Unknown)),
+
+					_ => {
+						if self.tydecls.get(name).is_none() {
+							self.report(Report::AnalyzeUnknownTypeRef {
+								span: ref_ty.span(),
+								name,
+							});
+						}
+
+						Ty::Ref(name)
 					}
-
-					Ty::Ref(name)
 				}
-			},
+			}
+
+			SyntaxTyKind::Enum(enum_ty) => Ty::Enum(self.enum_ty(enum_ty)),
+
+			SyntaxTyKind::Struct(struct_ty) => Ty::Struct(self.struct_ty(struct_ty)),
 
-			SyntaxTyKind::Enum(syntax_enum) => Ty::Enum(syntax_enum.into_config(state)),
-			SyntaxTyKind::Struct(syntax_struct) => Ty::Struct(syntax_struct.into_config(state)),
-			SyntaxTyKind::Instance(name) => Ty::Instance(name.map(|name| name.into_config())),
+			SyntaxTyKind::Instance(instance_ty) => Ty::Instance(instance_ty.as_ref().map(|ty| ty.name)),
 		}
 	}
-}
 
-impl<'src> SyntaxEnum<'src> {
-	fn into_config(self, state: &mut ConvertState<'src>) -> Enum<'src> {
-		let span = self.span();
+	fn enum_ty(&mut self, ty: &SyntaxEnum<'src>) -> Enum<'src> {
+		let span = ty.span();
 
-		match self.kind {
+		match &ty.kind {
 			SyntaxEnumKind::Unit(enumerators) => {
 				if enumerators.is_empty() {
-					state.push_report(Report::AnalyzeEmptyEnum { span })
+					self.report(Report::AnalyzeEmptyEnum { span });
 				}
 
-				Enum::Unit(enumerators.into_iter().map(|v| v.into_config()).collect())
+				Enum::Unit(enumerators.iter().map(|e| e.name).collect())
 			}
 
 			SyntaxEnumKind::Tagged { tag, variants } => {
-				if variants.is_empty() {
-					state.push_report(Report::AnalyzeEmptyEnum { span })
-				}
-
-				let tag_name = tag.into_config();
+				let tag_name = tag.value;
+
+				let variants = variants
+					.iter()
+					.map(|variant| {
+						if variant.1.fields.iter().any(|field| field.0.name == tag_name) {
+							self.report(Report::AnalyzeEnumTagUsed {
+								tag_span: tag.span(),
+								used_span: variant.0.span(),
+								tag: tag_name,
+							});
+						}
+
+						(variant.0.name, self.struct_ty(&variant.1))
```

Comment:
Let's not use `varient.x` -> move to a named tuple such as `.map(|(key, value)|`